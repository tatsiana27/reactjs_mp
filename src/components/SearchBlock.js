import React, { Component } from 'react';
import { connect } from 'react-redux';
import isEmpty from 'lodash/isEmpty';
import { searchMovies, clearMovies } from '../actions/movies.action';
import { isValueSearchInputValue } from '../helpers/is-valid-search-input-value';
import { RadioButton } from './RadioButton';
import { FILTERS } from '../data/data';

import '../styles/search-block.scss';

export class SearchBlockUI extends Component {
    state = {
        search: '',
        searchBy: 'title',
        isValid: false
    };

    onSubmit = (event) => {
        event.preventDefault();

        const { search, searchBy, isValid } = this.state;
        const { searchMovies, clearMovies } = this.props;

        clearMovies();

        if((isEmpty(search) && !isValid) || isEmpty(searchBy)) {
            return;
        }

        searchMovies({ search, searchBy});
    };

    onChangeInput = (event) => {
        const { searchBy } = this.state;
        const { value: search } = event.target;

        this.setState({ search, searchBy, isValid: isValueSearchInputValue(search) });
    };

    onRadioChange = (event) => {
        this.setState({
            searchBy: event.currentTarget.value
        });
    };

    render() {
        const { search, searchBy, isValid } = this.state;

        return (
            <form method="#" action="#" className="search-block" onSubmit={this.onSubmit}>
                <h4 className="title">Find your movie</h4>
                <input type="text"
                       id="movie-title"
                       autoFocus
                       onChange={this.onChangeInput}
                       value={search}
                />
                <span className={isValid ? 'd-none' : 'invalid-input'}>Search value is not valid</span>

                <div className="filter-block">
                    <div>
                        <p>search by</p>
                        {FILTERS.map(filter =>
                            <RadioButton
                                key={filter.id}
                                name={filter.name}
                                id={filter.id}
                                value={filter.name}
                                checked={searchBy === filter.name}
                                onChange={this.onRadioChange}
                            />)
                        }
                    </div>
                    <div>
                        <button id="search-btn" type="submit">Search</button>
                    </div>
                </div>
            </form>
        )
    };
}

export const mapStateToProps = (state) => ({
    movies: state.movies,
});

export const mapDispatchToProps = (dispatch) => ({
    searchMovies: payload => searchMovies(dispatch, payload),
    clearMovies: payload => clearMovies(dispatch, null),
});

export const SearchBlock = connect(mapStateToProps, mapDispatchToProps)(SearchBlockUI);

