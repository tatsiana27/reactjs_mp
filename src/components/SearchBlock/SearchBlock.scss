@import '../../styles/variables';

.search-block {
  font-size: 12px;
  display: block;
  padding:  0 40px;
}
.search-block .title {
  text-transform: uppercase;
  color: $white;
}

#movie-title {
  width: 100%;
  color: $white;
  background-color: $black;
  border: 1px solid $black;
  border-bottom: 1px solid $light-coral;
  padding: 10px;
}

.filter-block {
  display: flex;
  justify-content: space-between;
  position: relative;
}

.filter-block div > p {
  text-transform: uppercase;
  display: inline-block;
  margin-right: 10px;
}
.filter-block {
  label {
    text-transform: uppercase;

    background-color: $light-gray;
    padding: 4px 20px;
    margin: {
      left: 4px;
      right: 4px;
    }
    cursor: pointer;
  }
  input[type="radio"]:checked + label {
    background-color: $light-coral;
  }
}

#search-btn {
  background-color: $light-coral;
  border: 1px solid $light-coral;
  padding: 2px 40px;
  text-transform: uppercase;
  color: $white;
  position: relative;
  top: 8px;
  right: -20px;
}

.d-none {
  display: none;
}
.invalid-input {
  color: $red1;
}

.info-block {
  color: $black;
  padding: {
    left: 80px;
    right: 80px;
    top: 15px;
    bottom: 15px;
  };
  position: absolute;
  bottom: -45px;
  left: 0;
  right: 0;
  display: flex;

  justify-content: space-between;
  background-color: $light-gray;
  .sorting-block {
    display: flex;
    p {
      margin: 0;
    }
    label {
      margin: 0 20px;
      cursor: pointer;
    }
  }
  input[type="radio"]:checked + label {
    color: $red;
  }
}
