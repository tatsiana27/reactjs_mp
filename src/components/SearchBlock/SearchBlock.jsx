import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import isEmpty from 'lodash/isEmpty';
import { bindActionCreators } from 'redux';
import { fetchMovies, updateParams } from '../../modules/movies';
import { isValidSearchInputValue } from '../../helpers/is-valid-search-input-value';
import { FILTERS, SORTINGS } from '../../data';

import { RadioButton } from '../RadioButton';
// import PropTypes from 'prop-types';
//
// import Movie from '../Movie';
// import EmptyResult from '../EmptyResult';
import './SearchBlock.scss';

export class SearchBlock extends Component {

  state = {
    search: '',
    searchBy: 'title',
    sortBy: 'raiting',
    isValid: false,
  };

  onSubmit = (event) => {
    event.preventDefault();

    const { search, searchBy, isValid } = this.state;
    const { fetchMovies, params } = this.props;

    if((isEmpty(search) && !isValid) || isEmpty(searchBy)) {
      return;
    }

    fetchMovies({ search: params.search, searchBy: params.searchBy, sortBy: params.sortBy });
  };

  onChangeInput = (event) => {
    const { value: search } = event.target;
    const { updateParams } = this.props;
    const { searchBy, sortBy } = this.state;

    this.setState({ search, searchBy, isValid: isValidSearchInputValue(search) });
    updateParams({search: search, searchBy: searchBy, sortBy: sortBy});
  };

  onSearchRadioChange = (event) => {
    const { value: searchBy } = event.currentTarget;
    const { updateParams } = this.props;

    this.setState({
      searchBy: event.currentTarget.value
    });

    const { search, sortBy } = this.state;
    updateParams({search: search, searchBy: searchBy, sortBy: sortBy});

  };

  onSortRadioChange = (event) => {
    const { value: sortBy } = event.currentTarget;
    const { updateParams } = this.props;

    this.setState({
      sortBy: event.currentTarget.value
    });

    const { search, searchBy } = this.state;
    updateParams({search: search, searchBy: searchBy, sortBy: sortBy});
  };

  render() {
    const { search, searchBy, sortBy, isValid } = this.state;
    const { movies } = this.props;

    return (
      <form method="#" action="#" className="search-block" onSubmit={this.onSubmit}>
        <h4 className="title">Find your movie</h4>
        <input type="text"
               id="movie-title"
               autoFocus
               onChange={this.onChangeInput}
               value={search}
        />
        <span className={isValid ? 'd-none' : 'invalid-input'}>Search value is not valid</span>

        <div className="filter-block">
          <div>
            <p>search by</p>
            {FILTERS.map(filter =>
              <RadioButton
                key={filter.id}
                name={filter.name}
                id={filter.id}
                value={filter.name}
                checked={searchBy === filter.name}
                onChange={this.onSearchRadioChange}
              />)
            }
          </div>
          <div>
            <button id="search-btn" type="submit">Search</button>
          </div>
        </div>
        <div className="info-block">
          <span className="movies-count">{movies.total ? movies.total : 0} movies found</span>
          <div className="sorting-block">
            <p>Sort by</p>
            {SORTINGS.map(item =>
              <RadioButton
                key={item.id}
                name={item.name}
                id={item.id}
                value={item.name}
                checked={sortBy === item.name}
                onChange={this.onSortRadioChange}
              />)
            }
          </div>
        </div>
      </form>
    )
  };
}

SearchBlock.propTypes = {
  movies: PropTypes.shape({}).isRequired,
};

const mapStateToProps = state => ({
  loading: state.movies.loading,
});
const mapDispatchToProps = dispatch => bindActionCreators({
  fetchMovies,
  updateParams,
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(SearchBlock);
