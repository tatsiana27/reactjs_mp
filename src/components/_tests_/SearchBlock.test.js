import React from 'react';
import { SearchBlock } from '../SearchBlock';
import { shallow } from 'enzyme';

describe('SearchBlock ', () => {
    test('should render SearchBlock Component', () => {
        const wrapper = shallow(
            <SearchBlock/>
        );

        expect(wrapper).toMatchSnapshot();
    });
});