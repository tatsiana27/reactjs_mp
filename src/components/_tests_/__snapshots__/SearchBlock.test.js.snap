// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBlock  should render SearchBlock Component 1`] = `
<form
  action="#"
  className="search-block"
  method="#"
>
  <h4
    className="title"
  >
    Find your movie
  </h4>
  <input
    id="movie-title"
    name="movie"
    type="text"
  />
  <div
    className="filter-block"
  >
    <div>
      <p>
        search by
      </p>
      <RadioButton
        id={1}
        isChecked={true}
        key="1"
        name="title"
        value="title"
      />
      <RadioButton
        id={2}
        isChecked={false}
        key="2"
        name="genre"
        value="genre"
      />
    </div>
    <div>
      <button
        id="search-btn"
        type="submit"
      >
        Search
      </button>
    </div>
  </div>
</form>
`;
